### ISSUE TITLE
Restrictions for iTunes categories

### ISSUE THREAD

**Issue by mpgirro** [mpgirro/stalla#43]
There are additional restriction to be applied for our iTunes category implementations:

- [x] Restrict to predefined values from [Apple Podcasts Categories](https://help.apple.com/itc/podcasts_connect/#/itc9267a2f12) (see PR #42)
- [ ] ~Migrate from using `List` of categories to `Set` since duplicates make no sense.~
- [ ] ~[Apple Podcaster’s Guide to RSS](https://help.apple.com/itc/podcasts_connect/#/itcb54353390) allows to add several categories (although in their directory, only the first will be used). They do not specify if it is valid to declare a category separately when there are also subcategories of the category declared. It does not make much sense nevertheless. We can simply substitute an existing `ItunesCategory.Simple` instance with an added `ItunesCategory.Nested` instances that reference the existing `Simple` to avoid such cases and enforce more unambiguous feeds.~

**Comment by rock3r** [comment]
Not sure what you mean with the last point, tbh, when you mention substituting `Simple` with `Nested` instances. Do you have an example handy?

**Comment by mpgirro** [comment]
Sorry, this was phrased very poorly. What I mean is the following situation:

```
...
<channel>
  ...
  <itunes:category text="News" />

  <itunes:category text="News">
      <itunes:category text="Tech News" />
  </itunes:category>
  ...
</channel>
```

The seconds nested entry implies the first. So the first entry is obsolete in terms of informational value it transports. Changing the collection from `List` to `Set` avoids redundant information, but only by comparing instances directly. We can enhance this by also taking implied information into account. When adding a `Nested` we can check if the parent `Simple` is already present. If it is, then the `Simple` is obsolete and can be removed from the set, since the `Nested` transport also the parent information. 

It just results in cleaner feeds.

**Comment by rock3r** [comment]
Ah I see what you mean. Well, while I see the reasoning, I can't say I agree with this. You'd be altering the information the feed carries — even by just using a Set, you're dropping information which is important. Let me elaborate:

1. As you noted correctly, the positional information for categories is important. On iTunes for example, we need to be able to preserve the information about which one is the first category, since that's what Apple uses at the end of the day. Using a `Set`, or anything but a `List` really, loses this crucially important piece of information. So I don't believe a `Set` is the right thing to use — not even a linked one, for the reasons I'll cover next
2. Not only the order in which categories appear is important (and may carry meaning even beyond what is in the first position), but also which ones appear is important. Doing consolidation automatically on behalf of library users can cause a lot of confusion and errors. I for one would be upset as a user if the library weren't faithfully serializing and deserializing data to and from an RSS feed.
3. Not to mention, everything in the code so far is made in a way as to make the model, parser, and writer as transparent as possible and able to represent exactly 1:1 the data in the RSS. This would be a drastic departure, and would also result in breaking the ability to read-write-read back data that we've kept in so far. In my head, any data manipulation that changes the meaning or representation should be done outside of the library, by our clients, which know what their use case is and whether such manipulations are appropriate for them.

In light of these, I'd argue here (and, similarly, for Play categories) we should keep using a `List`, so that we keep both positional information and original intent. I strongly believe we shouldn't alter data automatically. What we can do, should we strongly feel like there is a need, is to offer an API to, given a list of categories, consolidate it as you said. This way, the library users have the choice of doing this without having to write much code at all, but we don't risk surprising users who just want to read or write something "as is" — which is in my mind the most important use case.

So, what about creating a `ModelOptimizer` of sorts to perform the operations you mentioned? It would also go well in tandem with something else I was thinking about, which is a `ModelValidator` — of which I already had created a prototype before starting to work on Stalla, and that I would be happy to contribute to. That checks for empty values, maximum lenghts, value constraints, etc etc.

**Comment by mpgirro** [comment]
You make a very good argument against the changes. I see, I should have put this topic into a GH Discussion first, instead of just adding them to this issue. Sorry about that.

The order of appearance argument is totally correct and for itself should have already hindered me from bringing this change up at all. Transparent feed data representation only more so.

Besides that, I think it's a very good idea of providing the functionality in a `ModelOptimizer`. I also really like your idea of an `ModelValidator`. Coincidentally I also had an idea a while ago that fits into this category: providing some sort of evaluation tool that gives you an assessment of the "quality" (that would have to be specified first of course) of the feed (model) you are providing. These ideas need more brainstorming of course. Definitely something we can plan to work on!

**Comment by rock3r** [comment]
Ahah don't worry. I hadn't even thought about the discussions stuff. They kinda feel a bit weird to use for this stuff, more like Stackoverflow Q&A than an actual discussion.

The quality check seems something that would indeed go hand in hand with a validation engine. Probably worth splitting off to its own ticket tho. :)

**Comment by mpgirro** [comment]
Just pushed the KDoc table. This item is done now
